basic/src/test/java/org/hibernate/benchmark/enhancement/EntityInstantiators.java

```diff
@@ -0,0 +1,259 @@
+package org.hibernate.benchmark.enhancement;
+
+import java.lang.reflect.Method;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.hibernate.HibernateException;
+import org.hibernate.Session;
+import org.hibernate.SessionFactory;
+import org.hibernate.boot.registry.StandardServiceRegistryBuilder;
+import org.hibernate.bytecode.enhance.spi.EnhancementContext;
+import org.hibernate.bytecode.enhance.spi.Enhancer;
+import org.hibernate.bytecode.internal.none.BytecodeProviderImpl;
+import org.hibernate.bytecode.spi.BasicProxyFactory;
+import org.hibernate.bytecode.spi.BytecodeProvider;
+import org.hibernate.bytecode.spi.ProxyFactoryFactory;
+import org.hibernate.bytecode.spi.ReflectionOptimizer;
+import org.hibernate.cfg.AvailableSettings;
+import org.hibernate.cfg.Configuration;
+import org.hibernate.engine.spi.SessionFactoryImplementor;
+import org.hibernate.engine.spi.SharedSessionContractImplementor;
+import org.hibernate.metamodel.internal.EntityInstantiatorPojoOptimized;
+import org.hibernate.metamodel.internal.EntityInstantiatorPojoStandard;
+import org.hibernate.metamodel.spi.EntityInstantiator;
+import org.hibernate.property.access.spi.PropertyAccess;
+import org.hibernate.proxy.HibernateProxy;
+import org.hibernate.proxy.ProxyFactory;
+import org.hibernate.type.CompositeType;
+
+import jakarta.persistence.Entity;
+import jakarta.persistence.Id;
+import org.openjdk.jmh.annotations.AuxCounters;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+import org.openjdk.jmh.infra.Blackhole;
+
+/**
+ * @author Marco Belladelli
+ */
+@State( Scope.Thread )
+public class EntityInstantiators {
+	private static final int FORTUNES = 100;
```

Comment:
Not sure what's a good number of instances for each query here, tried also `1_000` and `10_000` to similar results.

Reply:
you can make it as a @Param in the JMH benchmark

Reply:
Sure, but since the results are basically the same (in terms of ops and flamegraphs) does it matter?

Reply:
Yep let me explain why: usually while allocating tons of stuff the JIT speed it up using a thing called TLABs (see https://shipilev.net/jvm/anatomy-quarks/4-tlab-allocation/ for more info): the sad reality of this TLABs machinery is that it can hit the limits of what your hardware can deliver (in term of cost of accessing the main memory or the cache L3 bandwidth): given that this bandwidth is a previous and limited resource and given that using that parameter can help tuning the allocation rate in order to NOT be the main bottleneck, this can be used to understand if the differential of performance (if any) is coming from the allocation strategy (reflective or not).
Otherwise, the risk, is to hit the memory bandwidth limit due to TLAB allocation and changing the allocation bandwidth won't make any difference because the main bottleneck is elsewhere.
I hope this will make more sense.

Reply:
Sure, I understand, but I'm not sure it is necessary in this case: testing with either 100 or 1000 entity instances basically produces the same results / difference in performance (around 2.5/3%), and we can see the impact of the reflection on the CPU cycles clearly from the flamegraph. I don't think fetching more entity instances per query makes a difference for instantiation, but maybe I didn't get the full picture.

Also, I already tested with `-prof gc` for allocation rate, you can see the data in the PR description (1223.515 MB/s for `optimized` vs 1190.467 MB/s for `standard`).

Reply:
Yep, this is likely the limit I was talking about - to understand if is the main limit, you can try this way:
make it parametric and run the test with 10 or 100 instances and compare the allocation rate for both; if it's the same it means that even with 10, the main limit is the memory bandwidth.
If is that so you have 2 options:
- lower the number in order to NOT making it the main bottleneck - but the risk is that it won't ever happen
- just pick the minimum value which use the same and, as you rightly say, look at the flamegraph to still understand the diff of cycles in the allocation path

For the latter, is not trivial to just use the cycles to attributes the cost of a method call (especially which allocates!), 
because pu profiling on modern CPU for nano-benchs (like these) is fairly imprecise and the actual cost often hit nearby methods (in the assembly - not something we can do about it, apart using so-called precise perf events in async profiler, which have some better accuracy) which willl be blamed instead.
Not saying that checking the cycles is useless but depends, which is the universal answer for everything :P

Reply:
The results for 100 instances are what I pasted above. Here are the results of an example run with 10 instances:

```
EntityInstantiators.optimized                      thrpt   10   304416.116 ±  3732.685   ops/s
EntityInstantiators.optimized:instances            thrpt   10  3044161.161 ± 37326.847   ops/s
EntityInstantiators.optimized:·async               thrpt               NaN                 ---
EntityInstantiators.optimized:·gc.alloc.rate       thrpt   10     1591.699 ±    15.612  MB/sec
EntityInstantiators.optimized:·gc.alloc.rate.norm  thrpt   10     5487.755 ±    44.622    B/op
EntityInstantiators.optimized:·gc.count            thrpt   10      438.000              counts
EntityInstantiators.optimized:·gc.time             thrpt   10      664.000                  ms
EntityInstantiators.standard                       thrpt   10   300179.743 ±  3772.560   ops/s
EntityInstantiators.standard:instances             thrpt   10  3001797.429 ± 37725.600   ops/s
EntityInstantiators.standard:·async                thrpt               NaN                 ---
EntityInstantiators.standard:·gc.alloc.rate        thrpt   10     1569.682 ±    18.574  MB/sec
EntityInstantiators.standard:·gc.alloc.rate.norm   thrpt   10     5487.630 ±     6.175    B/op
EntityInstantiators.standard:·gc.count             thrpt   10      513.000              counts
EntityInstantiators.standard:·gc.time              thrpt   10      729.000                  ms
```

There still is a ~2% increase in prformance in the optimized case, and the allocation rate this time is higher: `1591.699 MB/s optimized` - `1569.682 MB/s standard` with 10 instances vs `1223.515 MB/s optimized` - `1190.467 MB/s standard` with 100 instances.

Reply:
which is a good sign re the bottleneck (if we can allocate more - even if slightly more) it means that it wasn't a saturated resource  - now let me send a PR to your branch to add polymorphism or if I can write some proper comment to explain the next step

Reply:
Hey @franz1981 any news? Not sure what else is needed to prove the change on the Quarkus side is beneficial - I was hoping to merge that soon.

Reply:
Le's have a quick call in few minutes
I got something to show :) 
let's sync on slack ok?

Reply:
I've updated the PR with your changes to the benchmark, find the results in my comment below.